### PR TITLE
sci-libs/tensorflow: add 2.13.1, 2.14.1, 2.15.0

### PR DESCRIPTION
**Pull request by wangjiezhe**
Update tensorflow, keras, tensorflow-estimator and tensorboard to version 2.13.

Tested on RTX4060 and `dev-util/nvidia-cuda-toolkit-12.3.2` with `USE="cuda mpi xla"` and `TF_CUDA_COMPUTE_CAPABILITIES=8.9`.

Here is an explanation of dependency problem of tensorflow-2.13:

1. flatbuffers-23.1.21 is required by tensorflow-2.13, as shown in [tensorflow/lite/schema/schema_generated.h](https://github.com/tensorflow/tensorflow/blob/v2.13.1/tensorflow/lite/schema/schema_generated.h#L25).
2. tensorflow>=2.13 needs `absl/log/check.h`, as shown in [tensorflow/compiler/jit/xla_compile_on_demand_op.cc](https://github.com/tensorflow/tensorflow/blob/v2.13.1/tensorflow/compiler/jit/xla_compile_on_demand_op.cc#L25), which is introduced in [abseil-cpp-20230125](https://github.com/abseil/abseil-cpp/releases/tag/20230125.0).
3. **Tensorflow needs abseil-cpp to be built with C++17**, whereas [abseil-cpp>=20230125.3](https://github.com/gentoo/gentoo/blob/master/dev-cpp/abseil-cpp/abseil-cpp-20230125.3-r1.ebuild#L56) in the official repo is built with C++14, and here is the [discussion](https://github.com/gentoo/gentoo/pull/32281). (This is the main reason of the following dependency problems.)
4. Therefore, the only available version of abseil-cpp compatible to tensorflow now is [20230125.2](https://github.com/gentoo/gentoo/blob/master/dev-cpp/abseil-cpp/abseil-cpp-20230125.2.ebuild#L61).
5. net-libs/grpc is built with specific abseil-cpp version. So the eligible versions are 1.55.1 and 1.56.1.
6. net-libs/grpc-1.56.1 requires dev-libs/protobuf>=23.3, which requires dev-cpp/abseil-cpp>=20230125.3 and is shown in https://bugs.gentoo.org/912796.
7. Therefore, the only available version of grpc compatible to tensorflow now is 1.55.1.
8. net-libs/grpc-1.55.1 requires dev-libs/protobuf>=22, so protobuf-22.5 and protobuf-python-4.22.5 are added. Since abseil-cpp is built with c++17, protobuf and protobuf-python also require to be built with c++17, where they are built with c++14 by default (and in other versions).

In conclusion, the only successful dependencies to compile tensorflow-2.13 is:

- dev-cpp/abseil-cpp-20230125.2
- dev-libs/protobuf-22.5
- dev-python/protobuf-python-4.22.5
- net-libs/grpc-1.55.1
- dev-libs/flatbuffers-23.1.21
- dev-python/flatbuffers-23.1.21

Merge should be cautious. I don't know whether there are conflicts with other packages.

Closes: https://bugs.gentoo.org/881445
Closes: https://bugs.gentoo.org/882617
Closes: https://bugs.gentoo.org/905673
Closes: https://bugs.gentoo.org/913731
Closes: https://bugs.gentoo.org/915621
Closes: https://bugs.gentoo.org/906914
Closes: https://bugs.gentoo.org/917689
Closes: https://bugs.gentoo.org/918703

### CODE REVIEW

**Comment by SoapGentoo** [comment]
Honestly, I would prefer we put this in a `tensorflow` overlay, since I just don't see the breakage ending.

**Comment by perfinion** [comment]
Wow thanks! It'll take me a bit to review it properly, will have more comments during the weekend

File: dev-python/flatbuffers/flatbuffers-23.1.21.ebuild [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_USE_PEP517=setuptools
+PYTHON_COMPAT=( python3_{9..11} )
```
Comment (AndrewAmmerlaan) [inline dev-python/flatbuffers/flatbuffers-23.1.21.ebuild:7]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: dev-python/flatbuffers/flatbuffers-23.1.21.ebuild [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+"
+SRC_URI="
+	https://github.com/google/${PN}/archive/v${PV}.tar.gz
+		-> ${P}.gh.tar.gz
+"
+S=${WORKDIR}/${P}/python
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="amd64 arm arm64 ~riscv x86"
```
Comment (AndrewAmmerlaan) [inline dev-python/flatbuffers/flatbuffers-23.1.21.ebuild:24]:
```suggestion
KEYWORDS="~amd64 ~arm ~arm64 ~riscv ~x86"
```

File: dev-python/ml_dtypes/ml_dtypes-0.2.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+DEPEND="
+	dev-python/numpy[${PYTHON_USEDEP}]
+"
+BDEPEND="
+	dev-python/pybind11[${PYTHON_USEDEP}]
+"
+
+python_prepare_all() {
+	rmdir third_party/eigen
+	cp -r "${WORKDIR}/eigen-${EIGEN_CommitId}" third_party/eigen
```
Comment (AndrewAmmerlaan) [inline dev-python/ml_dtypes/ml_dtypes-0.2.0.ebuild:33]:
```suggestion
	rmdir third_party/eigen || die
	cp -r "${WORKDIR}/eigen-${EIGEN_CommitId}" third_party/eigen || die
```

File: dev-python/flatbuffers/flatbuffers-23.1.21.ebuild [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	test? (
+		dev-python/numpy[${PYTHON_USEDEP}]
+	)
+"
+
+python_test() {
+	cd "${WORKDIR}/${P}/tests" || die
+	# zeroes means without benchmarks
+	"${EPYTHON}" py_test.py 0 0 0 false || die
+	"${EPYTHON}" py_flexbuffers_test.py -v || die
```
Comment (AndrewAmmerlaan) [inline dev-python/flatbuffers/flatbuffers-23.1.21.ebuild:38]:
```suggestion
	"${EPYTHON}" py_test.py 0 0 0 false || die "Tests failed with ${EPYTHON}"
	"${EPYTHON}" py_flexbuffers_test.py -v || die "Tests failed with ${EPYTHON}"
```
Its very useful to know which version of python the tests failed with.

File: dev-python/ml_dtypes/ml_dtypes-0.3.2.ebuild [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+DEPEND="
+	dev-python/numpy[${PYTHON_USEDEP}]
+"
+BDEPEND="
+	dev-python/pybind11[${PYTHON_USEDEP}]
+"
+
+python_prepare_all() {
+	rmdir third_party/eigen
+	cp -r "${WORKDIR}/eigen-${EIGEN_CommitId}" third_party/eigen
```
Comment (AndrewAmmerlaan) [inline dev-python/ml_dtypes/ml_dtypes-0.3.2.ebuild:33]:
```suggestion
	rmdir third_party/eigen || die
	cp -r "${WORKDIR}/eigen-${EIGEN_CommitId}" third_party/eigen || die
```

File: dev-python/protobuf-python/protobuf-python-4.22.5.ebuild [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	https://pypi.org/project/protobuf/
+"
+
+LICENSE="BSD"
+SLOT="0/22.5.0"
+
+S="${WORKDIR}/${PARENT_P}/python"
+
+BDEPEND="
+"
```
Comment (AndrewAmmerlaan) [inline dev-python/protobuf-python/protobuf-python-4.22.5.ebuild:42]:
```suggestion
```

File: dev-python/protobuf-python/protobuf-python-4.22.5.ebuild [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+DEPEND="
+	${PYTHON_DEPS}
+"
+RDEPEND="
+	${BDEPEND}
+	dev-libs/protobuf:${SLOT}
```
Comment (AndrewAmmerlaan) [inline dev-python/protobuf-python/protobuf-python-4.22.5.ebuild:48]:
```suggestion
	dev-libs/protobuf:${SLOT}
```

File: sci-libs/keras/keras-2.13.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,85 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/keras/keras-2.13.1.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/keras/keras-2.13.1.ebuild [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+inherit bazel distutils-r1
+
+DESCRIPTION="Deep Learning for humans"
+HOMEPAGE="https://keras.io/"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-libs/keras/keras-2.13.1.ebuild:15]:
```suggestion
```

File: sci-libs/keras/keras-2.14.0.ebuild [hunk 10]
```diff
@@ -0,0 +1,85 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/keras/keras-2.14.0.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/keras/keras-2.14.0.ebuild [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+inherit bazel distutils-r1
+
+DESCRIPTION="Deep Learning for humans"
+HOMEPAGE="https://keras.io/"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-libs/keras/keras-2.14.0.ebuild:15]:
```suggestion
```

File: sci-libs/keras/keras-2.15.0.ebuild [hunk 12]
```diff
@@ -0,0 +1,85 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/keras/keras-2.15.0.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..111} )
```

File: sci-libs/keras/keras-2.15.0.ebuild [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+inherit bazel distutils-r1
+
+DESCRIPTION="Deep Learning for humans"
+HOMEPAGE="https://keras.io/"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-libs/keras/keras-2.15.0.ebuild:15]:
```suggestion
```

File: sci-libs/tensorflow-estimator/tensorflow-estimator-2.13.0.ebuild [hunk 14]
```diff
@@ -0,0 +1,76 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow-estimator/tensorflow-estimator-2.13.0.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/tensorflow-estimator/tensorflow-estimator-2.13.0.ebuild [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+inherit bazel distutils-r1
+
+DESCRIPTION="A high-level TensorFlow API that greatly simplifies machine learning programming"
+HOMEPAGE="https://www.tensorflow.org/"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow-estimator/tensorflow-estimator-2.13.0.ebuild:19]:
```suggestion
```

File: sci-libs/tensorflow-estimator/tensorflow-estimator-2.14.0.ebuild [hunk 16]
```diff
@@ -0,0 +1,76 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow-estimator/tensorflow-estimator-2.14.0.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/tensorflow-estimator/tensorflow-estimator-2.14.0.ebuild [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+inherit bazel distutils-r1
+
+DESCRIPTION="A high-level TensorFlow API that greatly simplifies machine learning programming"
+HOMEPAGE="https://www.tensorflow.org/"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow-estimator/tensorflow-estimator-2.14.0.ebuild:19]:
```suggestion
```

File: sci-libs/tensorflow-estimator/tensorflow-estimator-2.15.0.ebuild [hunk 18]
```diff
@@ -0,0 +1,81 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow-estimator/tensorflow-estimator-2.15.0.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/tensorflow-estimator/tensorflow-estimator-2.15.0.ebuild [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+inherit bazel distutils-r1
+
+DESCRIPTION="A high-level TensorFlow API that greatly simplifies machine learning programming"
+HOMEPAGE="https://www.tensorflow.org/"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow-estimator/tensorflow-estimator-2.15.0.ebuild:19]:
```suggestion
```

File: sci-libs/tensorflow/tensorflow-2.13.1.ebuild [hunk 20]
```diff
@@ -0,0 +1,453 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_OPTIONAL=1
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow/tensorflow-2.13.1.ebuild:7]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/tensorflow/tensorflow-2.14.1.ebuild [hunk 21]
```diff
@@ -0,0 +1,446 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_OPTIONAL=1
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow/tensorflow-2.14.1.ebuild:7]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-libs/tensorflow/tensorflow-2.13.1.ebuild [hunk 22]
```diff
@@ -0,0 +1,453 @@
...
+		=sci-libs/tensorflow-estimator-${DEP_VER}*[${PYTHON_USEDEP}]
+	)"
+BDEPEND="
+	app-arch/unzip
+	=dev-build/bazel-5*
+	dev-java/java-config
+	cuda? (
+		>=dev-util/nvidia-cuda-toolkit-9.1[profiler]
+	)
+	!python? ( dev-lang/python )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow/tensorflow-2.13.1.ebuild:166]: This looks weird. It is almost always wrong to depend on python directly instead of `${PYTHON_DEPS}`

File: sci-libs/tensorflow/tensorflow-2.15.0.ebuild [hunk 23]
```diff
@@ -0,0 +1,464 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_OPTIONAL=1
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-libs/tensorflow/tensorflow-2.15.0.ebuild:7]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-visualization/tensorboard/tensorboard-2.13.0.ebuild [hunk 24]
```diff
@@ -0,0 +1,72 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-visualization/tensorboard/tensorboard-2.13.0.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-visualization/tensorboard/tensorboard-2.14.1.ebuild [hunk 25]
```diff
@@ -0,0 +1,72 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-visualization/tensorboard/tensorboard-2.14.1.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-visualization/tensorboard/tensorboard-2.14.1.ebuild [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+DESCRIPTION="TensorFlow's Visualization Toolkit"
+HOMEPAGE="https://www.tensorflow.org/"
+SRC_URI="$(pypi_wheel_url --unpack)"
+S=${WORKDIR}
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-visualization/tensorboard/tensorboard-2.14.1.ebuild:17]:
```suggestion
```

File: sci-visualization/tensorboard/tensorboard-2.15.1.ebuild [hunk 27]
```diff
@@ -0,0 +1,72 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_10 python3_11 )
```
Comment (AndrewAmmerlaan) [inline sci-visualization/tensorboard/tensorboard-2.15.1.ebuild:6]:
```suggestion
PYTHON_COMPAT=( python3_{10..11} )
```

File: sci-visualization/tensorboard/tensorboard-2.15.1.ebuild [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+DESCRIPTION="TensorFlow's Visualization Toolkit"
+HOMEPAGE="https://www.tensorflow.org/"
+SRC_URI="$(pypi_wheel_url --unpack)"
+S=${WORKDIR}
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE=""
```
Comment (AndrewAmmerlaan) [inline sci-visualization/tensorboard/tensorboard-2.15.1.ebuild:17]:
```suggestion
```

File: dev-python/flatbuffers/flatbuffers-23.1.21.ebuild (hunk 2 above)
Comment (wangjiezhe) [reply dev-python/flatbuffers/flatbuffers-23.1.21.ebuild:24]: This version has been stabilized before: https://bugs.gentoo.org/893044.
I'm not sure if I should add it back as unstable or not.

Comment (AndrewAmmerlaan) [reply dev-python/flatbuffers/flatbuffers-23.1.21.ebuild:24]: I think it's fine to add the stable keywords in that case.

**Comment by perfinion** [comment]
Merged in 77b20ee55bb5699ecd496e3a68e27fb4d5399f03 
Thanks!
